### PR TITLE
[MRG] ENH/MAINT: improvements to plot_topomap for custom outlines and image masks.

### PR DESCRIPTION
**Pull request by dengemann**
Allows us to generate figures like this one:

https://www.dropbox.com/s/5qpykqyh8fcc6bp/Screenshot%202015-04-17%2012.47.52.png?dl=0

### CODE REVIEW

**Comment by dengemann** [comment]
Ready for review @agramfort @mainakjas @Eric89GXL @fraimondo

**Comment by dengemann** [comment]
figure_3

**Comment by mainakjas** [comment]
@dengemann something is wrong. The tests fail.

```sh
======================================================================

ERROR: Test plotting epochs psd (+topomap)

----------------------------------------------------------------------

Traceback (most recent call last):

File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest

self.test(*self.arg)

File "/home/travis/build/mne-tools/mne-python/mne/viz/tests/test_epochs.py", line 131, in test_plot_psd_epochs

bands=[(0, 0.01, 'foo')]) # no freqs in range

File "/home/travis/miniconda/envs/testenv/lib/python2.7/unittest/case.py", line 473, in assertRaises

callableObj(*args, **kwargs)

File "/home/travis/build/mne-tools/mne-python/mne/epochs.py", line 645, in plot_psd_topomap

cbar_fmt=cbar_fmt, outlines=outlines, verbose=None)

TypeError: plot_epochs_psd_topomap() got an unexpected keyword argument 'outlines'
```

**Comment by dengemann** [comment]
@mainakjas fix is coming.

**Comment by dengemann** [comment]
Comments, any one? Quick review? I'm happy +1 for merge.

File: mne/epochs.py [hunk 1]
```diff
@@ -621,6 +621,15 @@ def plot_psd_topomap(self, bands=None, vmin=None, vmax=None, proj=False,
             False.
         cbar_fmt : str
             The colorbar format. Defaults to '%0.3f'.
+        outlines : 'head' | dict | None
+            The outlines to be drawn. If 'head', a head scheme will be drawn.
+            If dict, each key refers to a tuple of x and y positions.
+            The values in 'mask_pos' will serve as image mask. If None, nothing
+            will be drawn. Defaults to 'head'. If dict, the 'autoshrink' (bool)
+            field will, trigger automated shrinking of the positions due to
```
Comment (Eric89GXL) [inline mne/epochs.py]: no comma after "will"

File: mne/viz/topomap.py [hunk 2]
```diff
@@ -418,10 +405,10 @@ def plot_topomap(data, pos, vmax=None, vmin=None, cmap='RdBu_r', sensors=True,
...
         ylim = np.inf, -np.inf,
         mask_ = np.c_[outlines['mask_pos']]
-        xmin, xmax = (np.min(np.r_[xlim[0], mask_[:, 0] * 1.01]),
-                      np.max(np.r_[xlim[1], mask_[:, 0] * 1.01]))
-        ymin, ymax = (np.min(np.r_[ylim[0], mask_[:, 1] * 1.01]),
-                      np.max(np.r_[ylim[1], mask_[:, 1] * 1.01]))
+        xmin, xmax = (np.min(np.r_[xlim[0], mask_[:, 0]]),
+                      np.max(np.r_[xlim[1], mask_[:, 0]]))
+        ymin, ymax = (np.min(np.r_[ylim[0], mask_[:, 1]]),
+                      np.max(np.r_[ylim[1], mask_[:, 1]]))
```
Comment (Eric89GXL) [inline mne/viz/topomap.py]: BTW it would be more efficient to do a min of two mins than a min of a concat, but it doesn't really matter here

**Comment by Eric89GXL** [comment]
Can you add some simple tests of the functionality at least that the new input options actually run?

**Comment by Eric89GXL** [comment]
It would be nice if our coverage continued to go up instead of dwindling

**Comment by dengemann** [comment]
@Eric89GXL comments addressed.

**Comment by Eric89GXL** [comment]
Not quite, scroll up and look again https://github.com/mne-tools/mne-python/pull/1972#discussion-diff-28599004

**Comment by teonlamont** [comment]
whatever happened to coveralls?

**Comment by Eric89GXL** [comment]
We still have it running. Comments were disabled but I just reenabled them, so we can more clearly see the effect of PRs

File: mne/viz/topomap.py [hunk 3]
```diff
@@ -1249,7 +1298,15 @@ def plot_psds_topomap(
...
-
+    outlines : 'head' | dict | None
+        The outlines to be drawn. If 'head', a head scheme will be drawn.
+        If dict, each key refers to a tuple of x and y positions.
+        The values in 'mask_pos' will serve as image mask. If None, nothing
+        will be drawn. Defaults to 'head'. If dict, the 'autoshrink' (bool)
+        field will, trigger automated shrinking of the positions due to
+        points outside the outline. Moreover, a matplotlib patch object can
+        be passed for advanced masking options, either directly or as a
+        function that returns patches (required for multi-axis plots).
```
Comment (agramfort) [inline mne/viz/topomap.py]: missing empty line

**Comment by agramfort** [comment]
merge if you're happy and addressed comments. Disclaimer I did not test...

**Comment by agramfort** [comment]
@Eric89GXL or @mainakjas merge when you're happy

**Comment by Eric89GXL** [comment]
Will do, just waiting on some final tweaks, but @dengemann has disappeared. I remember back when he loved contributing to mne-python, perhaps he's moved on from for us now. I guess having the title of Dr. really changes some people... ;)

**Comment by agramfort** [comment]
LOL :)

**Comment by dengemann** [comment]
@Eric89GXL @agramfort I'm still in recovery mode, don't you be that mean ;)

**Comment by Eric89GXL** [comment]
Coding and code review should help you recover from... just about anything, really

**Comment by dengemann** [comment]
@Eric89GXL I did not quite get what remained unaddressed.

**Comment by dengemann** [comment]
Ah, ok, found it.

**Comment by dengemann** [comment]
I think now we're done.

**Comment by Eric89GXL** [comment]
Cool, will merge when green. Welcome back :)

**Comment by dengemann** [comment]
> Welcome back :)

Hah-hah! Wait until we meet next time ;)

**Comment by Eric89GXL** [comment]
Thanks @dengemann

**Comment by dengemann** [comment]
I'm glad we now have it on board, thanks for the review!

**Pull request merged**